Mathlib/CategoryTheory/Monoidal/Coherence.lean

```diff
@@ -0,0 +1,393 @@
+  `(tactic| monoidal_coherence)
+
+-- Porting note: restore this when `category_theory.bicategory.coherence` is ported.
+-- macro (name := pure_coherence') "pure_coherence" : tactic =>
+--   `(tactic| bicategory_coherence)
+
+example (X₁ X₂ : C) :
+    ((λ_ (𝟙_ C)).inv ⊗ 𝟙 (X₁ ⊗ X₂)) ≫ (α_ (𝟙_ C) (𝟙_ C) (X₁ ⊗ X₂)).hom ≫
+      (𝟙 (𝟙_ C) ⊗ (α_ (𝟙_ C) X₁ X₂).inv) =
+    𝟙 (𝟙_ C) ⊗ ((λ_ X₁).inv ⊗ 𝟙 X₂) := by
+  pure_coherence
+  -- This is just running:
+  -- change projectMap id _ _ (LiftHom.lift (((λ_ (𝟙_ C)).inv ⊗ 𝟙 (X₁ ⊗ X₂)) ≫
+  --     (α_ (𝟙_ C) (𝟙_ C) (X₁ ⊗ X₂)).hom ≫ (𝟙 (𝟙_ C) ⊗ (α_ (𝟙_ C) X₁ X₂).inv))) =
+  --   projectMap id _ _ (LiftHom.lift (𝟙 (𝟙_ C) ⊗ ((λ_ X₁).inv ⊗ 𝟙 X₂)))
+  -- exact congrArg _ (Subsingleton.elim _ _)
+
+namespace coherence
+
+/--
+Auxiliary simp lemma for the `coherence` tactic:
+this moves brackets to the left in order to expose a maximal prefix
+built out of unitors and associators.
+-/
+-- We have unused typeclass arguments here.
+-- They are intentional, to ensure that `simp only [assoc_LiftHom]` only left associates
+-- monoidal structural morphisms.
+@[nolint unusedArguments]
+lemma assoc_LiftHom {W X Y Z : C} [LiftObj W] [LiftObj X] [LiftObj Y]
+  (f : W ⟶ X) (g : X ⟶ Y) (h : Y ⟶ Z) [LiftHom f] [LiftHom g] :
+  f ≫ (g ≫ h) = (f ≫ g) ≫ h :=
+(Category.assoc _ _ _).symm
+
+/--
+Internal tactic used in `coherence`.
+
+Rewrites an equation `f = g` as `f₀ ≫ f₁ = g₀ ≫ g₁`,
+where `f₀` and `g₀` are maximal prefixes of `f` and `g` (possibly after reassociating)
+which are "liftable" (i.e. expressible as compositions of unitors and associators).
+-/
+macro (name := liftable_prefixes) "liftable_prefixes" : tactic => do
+  -- TODO we used to set the max instance search depth higher. Is this still needed?
+  `(tactic| (
+    simp only [monoidalComp, Category.assoc, MonoidalCoherence.hom] <;>
+    (apply (cancel_epi (𝟙 _)).1 <;> try infer_instance) <;>
+    -- TODO add `Bicategory.Coherence.assoc_LiftHom₂` when
+    -- `category_theory.bicategory.coherence` is ported.
+    simp only [assoc_LiftHom]))
+
+example {Y Z : C} (f : Y ⟶ Z) (g) (w : false) : (λ_ _).hom ≫ f = g := by
+  liftable_prefixes
+  guard_target = (𝟙 _ ≫ (λ_ _).hom) ≫ f = (𝟙 _) ≫ g
+  cases w
+
+lemma insert_id_lhs {C : Type _} [Category C] {X Y : C} (f g : X ⟶ Y) (w : f ≫ 𝟙 _ = g) : f = g :=
+by simpa using w
+lemma insert_id_rhs {C : Type _} [Category C] {X Y : C} (f g : X ⟶ Y) (w : f = g ≫ 𝟙 _) : f = g :=
+by simpa using w
+
+end coherence
+
+open coherence
+
+/-- If either the lhs or rhs is not a composition, compose it on the right with an identity. -/
+def insertTrailingIds : TacticM Unit := do
+  let some (_, lhs, rhs) := (← getMainTarget).eq? | throwError "Not an equality."
+  match lhs.getAppFnArgs with
+  | (``CategoryStruct.comp, _) => pure ()
+  | _ => evalTactic (← `(tactic| apply insert_id_lhs))
+  match rhs.getAppFnArgs with
+  | (``CategoryStruct.comp, _) => pure ()
+  | _ => evalTactic (← `(tactic| apply insert_id_rhs))
```

Comment:
I think this should be `def insertTrailingIds (mvarId : MVarId) : MetaM (List MVarId)` with the mvarId version of `apply` and then use `liftMetaTactic`

Reply:
Sure!